sled-agent/src/instance.rs

```diff
@@ -867,64 +938,188 @@ impl InstanceRunner {
+                        backend_id: SpecKey::Uuid(nic.id),
+                        interface_id: nic.id,
+                        // Add 8 to the NIC slot number to match Propolis's
+                        // previous interpretation of NIC slot parameters.
+                        pci_path: PciPath::new(0, nic.slot + 8, 0)?,
+                    },
+                    backend: VirtioNetworkBackend {
+                        vnic_name: port.name().to_string(),
+                    },
+                })
             })
-            .collect::<Result<Vec<_>, _>>()?;
-
-        let migrate = match migrate {
-            Some(params) => {
-                let migration_id = self.state
-                    .migration_in()
-                    // TODO(eliza): This is a bit of an unfortunate dance: the
-                    // initial instance-ensure-registered request is what sends
-                    // the migration ID, but it's the subsequent
-                    // instance-ensure-state request (which we're handling here)
-                    // that includes migration the source VMM's UUID and IP
-                    // address. Because the API currently splits the migration
-                    // IDs between the instance-ensure-registered and
-                    // instance-ensure-state requests, we have to stash the
-                    // migration ID in an `Option` and `expect()` it here,
-                    // panicking if we get an instance-ensure-state request with
-                    // a source Propolis ID if the instance wasn't registered
-                    // with a migration in ID.
-                    //
-                    // This is kind of a shame. Eventually, we should consider
-                    // reworking the API ensure-state request contains the
-                    // migration ID, and we don't have to unwrap here. See:
-                    // https://github.com/oxidecomputer/omicron/issues/6073
-                    .expect("if we have migration target params, we should also have a migration in")
-                    .migration_id;
-                Some(propolis_client::types::InstanceMigrateInitiateRequest {
-                    src_addr: params.src_propolis_addr.to_string(),
-                    src_uuid: params.src_propolis_id,
-                    migration_id,
+            .collect::<Result<Vec<NicComponents>, _>>()?;
+
+        let request = if let Some(params) = migrate {
+            // TODO(#6073): The migration ID is sent in the VMM registration
+            // request and isn't part of the migration target params (despite
+            // being known when the migration-start request is sent). If it were
+            // sent here it would no longer be necessary to read the ID back
+            // from the saved VMM/instance state.
+            let migration_id = self
+                .state
+                .migration_in()
+                .expect("migration ensure request entails migration in")
```

Comment:
Is this state something that's derived from the database, and therefore would this expect always fail due to some database state? Or is this state that is created a runtime? I'm worried this could cause a crash loop due to some database state.

Reply:
The previous comment attempted to explain this a bit; it is part of the migration saga behavior and I believe things would have to have gone *very* wrong for this to ever happen. I think we had a discussion about it at some point during #5749 and felt that the expect was okay, but I haven't been able to actually track that down...